core/src/main/java/tc/oc/pgm/stats/StatsModule.java

```diff
@@ -0,0 +1,90 @@
+package tc.oc.pgm.stats;
+
+import com.google.common.collect.ImmutableList;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.logging.Logger;
+import net.kyori.adventure.text.Component;
+import org.jdom2.Document;
+import org.jdom2.Element;
+import org.jetbrains.annotations.Nullable;
+import tc.oc.pgm.api.map.MapModule;
+import tc.oc.pgm.api.map.factory.MapFactory;
+import tc.oc.pgm.api.map.factory.MapModuleFactory;
+import tc.oc.pgm.api.match.Match;
+import tc.oc.pgm.api.match.MatchModule;
+import tc.oc.pgm.api.module.exception.ModuleLoadException;
+import tc.oc.pgm.api.player.MatchPlayer;
+import tc.oc.pgm.util.xml.InvalidXMLException;
+import tc.oc.pgm.util.xml.XMLFluentParser;
+import tc.oc.pgm.util.xml.XMLUtils;
+import tc.oc.pgm.variables.Variable;
+import tc.oc.pgm.variables.VariablesMatchModule;
+
+public class StatsModule implements MapModule<StatsMatchModule> {
+  private final List<MapVariableStat> variables;
+
+  private StatsModule(List<MapVariableStat> variables) {
+    this.variables = variables;
+  }
+
+  @Override
+  public @Nullable StatsMatchModule createMatchModule(Match match) throws ModuleLoadException {
+    final List<StatsMatchModule.VariableStat> variableStats = new ArrayList<>();
+    for (MapVariableStat mapVar : variables) {
+      if (mapVar.variable instanceof Variable.TrackedHolders<MatchPlayer>
+          && mapVar.variable instanceof Variable.DefinesDefault<MatchPlayer>) {
+        variableStats.add(new StatsMatchModule.VariableStat(mapVar.variable, mapVar.displayName));
+      }
+    }
+    return new StatsMatchModule(match, variableStats);
+  }
+
+  @Override
+  public Collection<Class<? extends MatchModule>> getSoftDependencies() {
+    return ImmutableList.of(VariablesMatchModule.class);
+  }
+
+  public static class Factory implements MapModuleFactory<StatsModule> {
+    @Override
+    public StatsModule parse(MapFactory factory, Logger logger, Document doc)
+        throws InvalidXMLException {
+      XMLFluentParser parser = factory.getParser();
+      Element statsElement = XMLUtils.getUniqueChild(doc.getRootElement(), "stats");
+      if (statsElement == null) {
+        return new StatsModule(Collections.emptyList());
+      } else {
+        Element additionalStatsElement = XMLUtils.getUniqueChild(statsElement, "additions");
+        List<MapVariableStat> statVars = new ArrayList<>();
+        if (additionalStatsElement != null) {
+          for (Element el : XMLUtils.getChildren(additionalStatsElement, "stat")) {
+            if (el.getAttribute("var") == null) {
+              throw new InvalidXMLException(
+                  "Expected 'var' attribute on custom stat definition", el);
+            }
```

Comment:
you do not need this, calling `.required()` on `parser.variable(el, "var").scope(MatchPlayer.class).required();` will already do this logic for you

Reply:
Cool can remove